Source/Core/Core/IOS/DI/DI.cpp

```diff
@@ -30,93 +34,618 @@ void DI::DoState(PointerWrap& p)
+    ERROR_LOG(IOS_DI, "Unknown ioctl 0x%02x", request.request);
+    return std::make_optional(SECURITY_ERROR);
+  }
+}
+
+std::optional<DI::DIResult> DI::StartDMATransfer(u32 command_0, u32 command_1, u32 command_2,
+                                                 u32 command_length, const IOCtlRequest& request)
+{
+  if (request.buffer_out_size < command_length)
+  {
+    // Actual /dev/di will still send a command, but won't write the length or output address,
+    // causing it to eventually time out after 15 seconds.  Just immediately time out here,
+    // instead.
+    WARN_LOG(IOS_DI,
+             "Output buffer is too small for the result of the command (%d bytes given, needed at "
+             "least %d); returning read timed out (immediately, instead of waiting)",
+             request.buffer_out_size, command_length);
+    return std::make_optional(READ_TIMED_OUT);
+  }
+
+  if ((command_length & 0x1f) != 0 || (request.buffer_out & 0x1f) != 0)
+  {
+    // In most cases, the actual DI driver just hangs for unaligned data, but let's be a bit more
+    // gentle.
+    WARN_LOG(IOS_DI,
+             "Output buffer or length is incorrectly aligned (buffer 0x%08x, buffer length %x, "
+             "command length %x)",
+             request.buffer_out, request.buffer_out_size, command_length);
+    return std::make_optional(BAD_ARGUMENT);
+  }
+
+  m_last_length = command_length;
+  DVDInterface::ExecuteCommand(command_0, command_1, command_2, request.buffer_out, command_length,
+                               true);
+  // Reply will be posted by done by FinishIOCtl.
+  return {};
+}
+std::optional<DI::DIResult> DI::StartImmediateTransfer(u32 command_0, u32 command_1, u32 command_2,
+                                                       const IOCtlRequest& request,
+                                                       bool write_to_buf)
+{
+  // TODO: handle write_to_buf being false (i.e. don't write back the value in DIIMMBUF)
+  if (write_to_buf && request.buffer_out_size < 4)
+  {
+    WARN_LOG(IOS_DI,
+             "Output buffer size is too small for an immediate transfer (%d bytes, should be at "
+             "least 4).  Performing transfer anyways.",
+             request.buffer_out_size);
+  }
+  // DVDInterface will handle writing directly (not accurate...)
+  DVDInterface::ExecuteCommand(command_0, command_1, command_2, request.buffer_out, 4, true);
+  // Reply will be posted by done by FinishIOCtl.
+  return {};
+}
+
+void DI::InterruptFromDVDInterface(DVDInterface::DIInterruptType interrupt_type)
+{
+  DIResult result;
+  switch (interrupt_type)
+  {
+  case DVDInterface::INT_TCINT:
+    result = SUCCESS;
+    break;
+  case DVDInterface::INT_DEINT:
+    result = DRIVE_ERROR;
+    break;
+  default:
+    PanicAlert("IOS::HLE::Device::DI: Unexpected DVDInterface interrupt %d!", interrupt_type);
+    result = DRIVE_ERROR;
+    break;
+  }
+
+  auto di = IOS::HLE::GetIOS()->GetDeviceByName("/dev/di");
+  if (di)
```

Comment:
Both statement bodies should be braced here, since one travels more than one line.